### PR TITLE
Add solver

### PR DESCRIPTION
**Pull request by aidangomez**

### CODE REVIEW

File: Source/Metal/NetBuffer.metal [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+using namespace metal;
+
+
+struct BufferDimensions {
+    ushort count;
+};
+
+kernel void reset_buffer(device float* input [[ buffer(0) ]],
+                            constant BufferDimensions& dims [[ buffer(1) ]],
```
Comment (alejandro-isaza) [inline Source/Metal/NetBuffer.metal]: alignment

Comment (aidangomez) [inline Source/Metal/NetBuffer.metal]: stupid xcode

File: Source/Solvers/SGDSolver.swift [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    public func updateSolverParameters() {
+        guard var params = params as? SGDSolverParameters else {
+            fatalError("Solver Parameters are incorrect type.")
+        }
+        params.learningRate = params.learningRate / (1 + params.momentum * Float(currentStep))
+    }
+
+    public func train(stepAction: (() -> Void)? = nil) {
+        done = true
```
Comment (alejandro-isaza) [inline Source/Solvers/SGDSolver.swift]: Do you mean false?

File: Source/Solvers/SGDSolver.swift [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            let commandBuffer = self.trainer.commandQueue.commandBuffer()
+            for n in net.nodes {
+                if let paramLayer = n.layer as? ParameterUpdateLayer {
+                    paramLayer.update(commandBuffer, updateFunction: self.updateFunction, solverParams: self.paramsBuffer)
+                }
+            }
+            commandBuffer.commit()
+            commandBuffer.waitUntilCompleted()
+
+            self.trainer.wipeBuffers()
```
Comment (alejandro-isaza) [inline Source/Solvers/SGDSolver.swift]: Why is this necessary? All operations overwrite the contents, don't they?

File: Tests/SGDSolverTests.swift [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+                return node.outputBuffer!.id
+            }
+            return val
+        }
+
+        let solver = try! SGDSolver(device: device, net: net, batchSize: batchSize, endStep: 1, learningRate: 0.5, momentum: 1.0)
+        let expecation1 = self.expectationWithDescription("Net forward pass 1")
+        let expecation2 = self.expectationWithDescription("Net forward pass 2")
+        solver.train({
+            solver.completion = { snapshot in
```
Comment (alejandro-isaza) [inline Tests/SGDSolverTests.swift]: a block inside a block? why?

File: Tests/SGDSolverTests.swift [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+                let sinkData = [Float](snapshot.forwardContentsOfBuffer(sinkBuffer))
```
Comment (alejandro-isaza) [inline Tests/SGDSolverTests.swift]: I realize now this should be called `contentsOfForwardBuffer`

File: Tests/SGDSolverTests.swift [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+                sink.consume(Blob(sinkData))
+
+                let sigmoid2Result = arrayFromBuffer(snapshot.forwardBuffers[sigmoid2BufferId])
```
Comment (alejandro-isaza) [inline Tests/SGDSolverTests.swift]: Use the public API

File: Tests/SGDSolverTests.swift [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+        net.connectBuffer(ip2Buffer, toLayer: ip2Layer)
+        net.connectLayer(ip2Layer, toBuffer: sigmoid2Buffer)
+        net.connectBuffer(sigmoid2Buffer, toLayer: sigmoid2Layer)
+        net.connectLayer(sigmoid2Layer, toBuffer: lossBuffer, atOffset: 0)
+        net.connectLayer(labelLayer, toBuffer: lossBuffer, atOffset: sigmoid2.outputSize)
+        net.connectBuffer(lossBuffer, toLayer: lossLayer)
+        net.connectLayer(lossLayer, toBuffer: sinkBuffer)
+        net.connectBuffer(sinkBuffer, toLayer: sinkLayer)
+
+        let sigmoid2BufferId = net.nodes.reduce(-1) { val, node in
```
Comment (alejandro-isaza) [inline Tests/SGDSolverTests.swift]: What is this?

File: Tests/SGDSolverTests.swift [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+                let lossResult = sink.data[0]
+                let lossExpected: Float = 0.298371109
+                XCTAssertEqualWithAccuracy(lossResult, lossExpected, accuracy: 0.0001)
+                expecation1.fulfill()
+
+                let net = solver.trainer.net
+                let commandBuffer = solver.trainer.commandQueue.commandBuffer()
+                for n in net.nodes {
+                    if let paramLayer = n.layer as? ParameterUpdateLayer {
+                        paramLayer.update(commandBuffer, updateFunction: solver.updateFunction, solverParams: solver.paramsBuffer)
```
Comment (alejandro-isaza) [inline Tests/SGDSolverTests.swift]: Why are you do doing this manually in the test?

File: Source/Layers/InnerProductLayer.swift [hunk 9]
```diff
@@ -68,7 +74,7 @@ public class InnerProductLayer: ForwardLayer, BackwardLayer {
         }
         biasesBuffer.label = "InnerProductBiases"
     }
-
+
```
Comment (alejandro-isaza) [inline Source/Layers/InnerProductLayer.swift]: 😬

File: Source/Layers/InnerProductLayer.swift [hunk 10]
```diff
@@ -139,4 +145,58 @@ public class InnerProductLayer: ForwardLayer, BackwardLayer {
...
+        do {
+            var weightDims = BufferDimensions(count: UInt32(weightsBuffer.length))
+            let weightDimsBuffer = buffer.device.newBufferWithBytes(&weightDims, length: sizeof(Parameters), options: .CPUCacheModeDefaultCache)
+            weightDimsBuffer.label = "InnerProductWeightsDimensions"
+
+            let encoder = buffer.computeCommandEncoder()
+            encoder.label = "InnerProductUpdateParamsWeights"
+            encoder.setComputePipelineState(updateFunction)
+            encoder.setBuffer(weightsBuffer, offset: 0, atIndex: 0)
```
Comment (alejandro-isaza) [inline Source/Layers/InnerProductLayer.swift]: This is very fragile to changes in the signature of updateFunction, and this will be repeated all over the place. Instead pass an object that takes care of this. Something like `update(updateAction: (MTLBuffer, MTLBuffer) -> Void) { updateAction(weightsBuffer, weightDiff) ... }`

File: Source/NetBuffer.swift [hunk 11]
```diff
@@ -5,6 +5,11 @@
 // contained in the file LICENSE at the root of the source code distribution
 // tree.
 
+
+struct BufferDimensions {
```
Comment (alejandro-isaza) [inline Source/NetBuffer.swift]: Not necessary for a single value.

File: Source/Metal/NetBuffer.metal [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+// contained in the file LICENSE at the root of the source code distribution
+// tree.
+
+#include <metal_stdlib>
+#include <metal_common>
+
+using namespace metal;
+
+
+struct BufferDimensions {
```
Comment (alejandro-isaza) [inline Source/Metal/NetBuffer.metal]: Not necessary for a single value.

File: Source/Layers/Layer.swift [hunk 13]
```diff
@@ -39,6 +39,11 @@ public protocol BackwardLayer: ForwardLayer {
     func encodeBackwardInBuffer(buffer: MTLCommandBuffer, batchSize: Int, outputDiff: MTLBuffer, input: MTLBuffer, inputDiff: MTLBuffer)
 }
 
+public protocol ParameterUpdateLayer {
```
Comment (alejandro-isaza) [inline Source/Layers/Layer.swift]: Now that I see what it does I think a better name is `TrainableLayer`

File: Source/Solvers/SGDSolver.swift (hunk 2 above)
Comment (aidangomez) [inline Source/Solvers/SGDSolver.swift]: No, I meant true. Maybe the variable name is misleading. @aleph7

File: Source/Solvers/SGDSolver.swift (hunk 3 above)
Comment (aidangomez) [inline Source/Solvers/SGDSolver.swift]: @aleph7 no. But, we could make it so they do? That might be the right approach. Assuming we view each layer's output as unique per call. (I.e. we wouldn't allow someone to write a layer that allows accumulation of output over calls.)

File: Tests/SGDSolverTests.swift (hunk 4 above)
Comment (aidangomez) [inline Tests/SGDSolverTests.swift]: @aleph7 because the `completion` in solver will wipe buffers before I have a chance to access. That's why I'm overwriting it with my assertions.
